AGDNS-4106-mv-articles-to-dns-kb (#439)
AdguardTeam/KnowledgeBaseDNS: docs/adguard-home/configuration.md

Reviewer: Add the header like the one other articles have, with the title and the sidebar position.

# Configuration

> [!WARNING]
> TODO(f.setrakov): !! Rework.
> Work on this article is still in progress. It may be reworked.

Most of these settings can be changed via the web-based admin interface. However, we decided to list them all here just in case.

> - [Command-line arguments](#command-line)

Reviewer 2: ~~We use an asterisk for bullet lists.~~

Well, I'm hallucinating.

- [Configuring upstreams](#upstreams)
    - [Specifying upstreams for domains](#upstreams-for-domains)
    - [Loading upstreams from file](#upstreams-from-file)
    - [Specifying upstreams for reverse DNS](#upstreams-for-rdns)
        - [Private addresses](#rdns-private)
        - [Public addresses](#rdns-public)
        - [Client lookups](#rdns-clients)
- [Configuration file](#configuration-file)
- [Reset web password](#password-reset)
- [Profiling with pprof](#pprof)

## Command-line arguments {#command-line}

Here is a list of all available command-line arguments.

> ``` bash

Reviewer: Please replace `bash` everywhere with a simple `sh`.

$ ./AdGuardHome --help
Usage:

./AdGuardHome [options]

> Options: -c, --config VALUE                 Path to the config file. -w, --work-dir VALUE               Path to the working directory. -h, --host VALUE                   Deprecated. Host address to bind HTTP server on. Use --web-addr. The short -h will work as --help in the future. -p, --port VALUE                   Deprecated. Port to serve HTTP pages on. Use --web-addr. --web-addr VALUE                   Address to serve the web UI on, in the host:port format. -s, --service VALUE                Service control action: status, install (as a service), uninstall (as a service), start, stop, restart, reload (configuration). -l, --logfile VALUE                Path to log file. If empty, write to stdout; if "syslog", write to system log. --pidfile VALUE                    Path to a file where PID is stored. --check-config                     Check configuration and exit. --no-check-update                  Don't check for updates. --update                           Update the current binary and restart the service in case it's installed. --no-mem-optimization              Deprecated. Disable memory optimization. --no-etc-hosts                     Deprecated: use clients.runtime_sources.hosts and dns.hostsfile_enabled instead. Do not use the OS-provided hosts. --local-frontend                   Use local frontend directories. -v, --verbose                      Enable verbose output. --glinet                           Run in GL-Inet compatibility mode. --no-permcheck                     Skip checking and migration of permissions of security-sensitive files. --version                          Show the version and exit. Show more detailed version description with -v. --help                             Print this help.

Reviewer 2: Bad formatting.

- `tcp://94.140.14.140`, `tcp://[2a10:50c0::1:ff]`: regular DNS (over TCP);

- `tcp://94.140.14.140:53`, `tcp://[2a10:50c0::1:ff]:53`: regular DNS (over TCP, with port).

- `tcp://dns-unfiltered.adguard.com`: regular DNS (over TCP, hostname);

- `tls://dns-unfiltered.adguard.com`: encrypted [DNS-over-TLS];

- `https://dns-unfiltered.adguard.com/dns-query`: encrypted [DNS-over-HTTPS];

- `h3://dns-unfiltered.adguard.com/dns-query`: encrypted [DNS-over-HTTPS] with forced [HTTP/3] and no fallback to HTTP/2 and below;

- `quic://dns-unfiltered.adguard.com`: encrypted [DNS-over-QUIC];

- `sdns://...`: [DNS Stamps] for [DNSCrypt]  or [DNS-over-HTTPS] resolvers;

- `[/example.local/]94.140.14.140`: DNS upstream for specific domains, see below;

- `[/*.example.local/]94.140.14.140`: DNS upstream for specific subdomains, see below;

### Specifying upstreams for domains {#upstreams-for-domains}

You can specify upstreams that will be used for specific domains using the dnsmasq-like syntax (see the documentation for the option `--server` [here][dnsmasq-man]). This feature is intended for private nameservers which deal with intranet domains.

The syntax is:

```none

> [/domain1/domain2/domainN/]upstreamString

Reviewer 2: I'd put the `upstreamString` in angle brackets to emphasise that it's a parameteric value, not a literal one.

```none
5.6.7.8:53
[//]4.3.2.1:53 [/com/]1.2.3.4:53
```

sends requests for `com` (and its subdomains) to `1.2.3.4:53`, requests for other top-level domains to `4.3.2.1:53`, and all other requests to `5.6.7.8:53`.

More specific domains take precedence over less specific domains. So, a configuration like this:

```none
[/host.com/]1.2.3.4 [/www.host.com/]2.3.4.5
```

sends queries for `*.host.com` to `1.2.3.4`, except for queries for `*.www.host.com`, which are sent to `2.3.4.5`.

The special server address `#` means “use the default servers”. So, a configuration like this:

```none
6.7.8.9
[/host.com/]1.2.3.4 [/www.host.com/]#
```

sends queries for `*.host.com` to `1.2.3.4` except for queries for `*.www.host.com`, which are sent to `6.7.8.9`, which is the default upstream.

Queries for the `DS` query type are following the assumption based on specification for records' presence given in [RFC 4035, section
2.4](https://datatracker.ietf.org/doc/html/rfc4035#section-2.4):

>  > A DS RRset SHOULD be present at a delegation point when the child zone is > signed. \[…\]  All DS RRsets in a zone MUST be signed, and DS RRsets MUST > NOT appear at a zone's apex.

Reviewer: Remove extra `>`s.

For example, the `DS` query for `domain.example.com` will be sent to the upstream specified for `example.com`, `com`, or the default one, even if there is a more specific upstream like `*.example.com`. Note that for two-label `DS` requests, the upstream specified for unqualified names, or the more specific one will be used, if any.

Wildcard `*` has a special meaning of "any subdomain", so `--upstream=[/*.host.com/]1.2.3.4` will send queries for `*.host.com` to `1.2.3.4`, but `host.com` will be forwarded to default upstreams.

#### Examples

- A configuration like:

    ```none
    8.8.8.8:53
    [/local/]192.168.0.1:53
    ```

>     sends queries for `*.local` domains to `192.168.0.1:53`. Other queries are

Reviewer: Here and below, single line.

    sent to `8.8.8.8:53`.

- A configuration like:

    ```none
    8.8.8.8:53
    [/host.com/]1.1.1.1:53
    [/maps.host.com/]#
    ```

    sends queries for `*.host.com` to `1.1.1.1:53` except for `*.maps.host.com`
    which are sent to `8.8.8.8:53` along with all other queries.

- A configuration like:

    ```none
    8.8.8.8:53
    [/host.com/]1.1.1.1:53
    [/*.host.com/]2.2.2.2:53
    ```

    sends queries for `*.host.com` to `2.2.2.2:53` except for `host.com` queries
    those are sent to `1.1.1.1:53`, but all other queries are sent to `8.8.8.8:53`.

- A configuration like:

    ```none
    8.8.8.8:53
    [/host.com/]1.1.1.1:53 2.2.2.2:53
    ```

    sends queries for `*.host.com` to `1.1.1.1:53` and `2.2.2.2:53`, but all
    other queries are sent to `8.8.8.8:53`.

### Loading upstreams from file {#upstreams-from-file}

Using specific upstreams for some domains is a common way to accelerate internet in China. For an example, see https://github.com/felixonmars/dnsmasq-china-list or any other of the many `dnsmasq` lists. These lists can be easily converted to a list for AdGuard Home:

```none

> Before:  server=/0-100.com/114.114.114.114 After:   [/0-100.com/]114.114.114.114

Reviewer 2: A line break woudn't hurt.  I'd even split the code block into two and move the words "Before" and "After" out of the blocks.

```

The problem with these lists is that they may be too large. In this case you may want to load them from a separate file instead of setting all upstreams in AdGuard Home settings. To do that, simply specify the path to a file with your list in the `upstream_dns_file` field of `AdGuardHome.yaml`.

> **BUG:** The file, just like the input in the web interface, currently doesn't accept internationalized domain names (e.g. `пример.рф` or `例子.中国`). As a workaround, convert them to Punycode (e.g. `xn--e1afmkfd.xn--p1ai` or `xn--fsqu00a.xn--fiqs8s` respectively). See [issue 2915][2915].

Reviewer 2: Use admonitions for all such notes.

```none
[/168.192.in-addr.arpa/]192.168.7.7
```

#### Client lookups {#rdns-clients}

AdGuard Home automatically gets the names of connected devices using reverse DNS lookup (rDNS). It sends `PTR` requests with the IP addresses of clients to appropriate DNS servers and uses the responses to enrich client information with human-friendly names.

This feature can be enabled and disabled with “Enable clients’ hostname resolution” setting in the “Upstream DNS servers” section or via the `clients.runtime_sources.rdns` field in the configuration file.

[2915]:           https://github.com/AdguardTeam/AdGuardHome/issues/2915
[DNS Stamps]:     https://dnscrypt.info/stamps/
[DNS-over-HTTPS]: https://en.wikipedia.org/wiki/DNS_over_HTTPS
[DNS-over-QUIC]:  https://datatracker.ietf.org/doc/html/rfc9250
[DNS-over-TLS]:   https://en.wikipedia.org/wiki/DNS_over_TLS
[DNSCrypt]:       https://dnscrypt.info/
[HTTP/3]:         https://en.wikipedia.org/wiki/HTTP/3
[dnsmasq-man]:    http://www.thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html
[private-ip]:     https://tools.ietf.org/html/rfc6303

> ## Configuration file {#configuration-fil}

Reviewer 2: This whole section needs to be reformatted according to our standards.

> ## Configuration file {#configuration-fil}

Reviewer: Who's Fil?

  Supports `http`, `https` and `socks5` schemes.
- `language`: UI language code, e.g. `en`.
- `theme`: The UI theme. The possible values are:
  `auto`, `dark`, `light`.
- `dns`: DNS configuration section.
    - **General settings**
        - `bind_hosts`: IP addresses on which to serve DNS queries. For each network interface there can only be one IP address of each IP version. Examples of valid configurations:

       ```yaml
       # Different network interfaces.
       'dns':
           'bind_hosts':
           - '127.0.0.1'
           - '192.168.1.1'
       ```

       ```yaml
       # Same network interface, different IP versions.
       'dns':
           'bind_hosts':
           - '127.0.0.1'
           - '::1'
       ```

       If you want your server to accept requests on all interfaces and using both IP versions, for example if you run a public server, put **one** item with the unspecified IP of any version:

       ```yaml
       'dns':
           'bind_hosts':
           - '0.0.0.0'
       ```

        - `port` — DNS server port to listen on.
        - `anonymize_client_ip` - If true, anonymize clients' IP addresses in logs and stats

>         - `blocked_services` (**before v0.107.37**): Blocked services settings

Reviewer: Remove outdated information.

---
title: Configuration

> sidebar_position: 7

Reviewer: Merge `master` and resync.  This position has already been taken by another article there.

---

:::warning

> Work on this article is still in progress. It may be reworked.

Reviewer 3: all our articles may be reworked, there is no need to add such an admonition here, I believe

Reviewer: @el-termikael, we are going to completely restructure this article in the future to bring it closer to what we have in e.g. AGDNS:

- https://github.com/AdguardTeam/AdGuardDNS/blob/master/doc/configuration.md

This is a reminder for us, and there will also be an additional Jira issue about that.

:::

This page contains descriptions of the command-line and configuration file settings.

:::note

Most of these settings can be changed via the web-based admin interface.

:::

## Command-line arguments {#command-line}

Here is a list of all available command-line arguments.

``` sh
$ ./AdGuardHome --help
Usage:

./AdGuardHome [options]

Options: 
-c, --config VALUE                 Path to the config file.
-w, --work-dir VALUE               Path to the working directory.
-h, --host VALUE                   Deprecated. Host address to bind HTTP server on. Use --web-addr. The short -h will work as --help in the future. 
-p, --port VALUE                   Deprecated. Port to serve HTTP pages on. Use --web-addr.
--web-addr VALUE                   Address to serve the web UI on, in the host:port format. 
-s, --service VALUE                Service control action: status, install (as a service), uninstall (as a service), start, stop, restart, reload (configuration).

> -l, --logfile VALUE                Path to log file. If empty, write to stdout; if "syslog", write to system log.

Reviewer 3: please use “double typographic quotes” here and below

Reviewer: @el-termikael, I agree in general, but this section is a verbatim output from the terminal, so it's important to keep the ASCII quotes here.

--pidfile VALUE                    Path to a file where PID is stored.
--check-config                     Check configuration and exit.
--no-check-update                  Don't check for updates.
--update                           Update the current binary and restart the service in case it's installed. 
--no-mem-optimization              Deprecated. Disable memory optimization.

> --no-etc-hosts                     Deprecated: use clients.runtime_sources.hosts and dns.hostsfile_enabled instead. Do not use the OS-provided hosts. 

Reviewer 3: clients.runtime_sources.hosts, dns.hostsfile_enabled, and stuff like that should be marked as code

Reviewer 3: also, after a colon, the next word should begin with a capital letter
see our editorial policy: https://app.notion.com/p/adguard/0df1a9bec9464fbf88bd813312ab0c26

Reviewer: @el-termikael

> clients.runtime_sources.hosts, dns.hostsfile_enabled, and stuff like that should be marked as code

See above, this is verbatim output from the terminal.

--local-frontend                   Use local frontend directories. 
-v, --verbose                      Enable verbose output. 
--glinet                           Run in GL-Inet compatibility mode.
--no-permcheck                     Skip checking and migration of permissions of security-sensitive files. 
--version                          Show the version and exit. Show more detailed version description with -v. 
--help                             Print this help.
```

> Please note, that the command-line arguments override settings from the configuration file.

Reviewer 3: ```suggestion
Please note that the command-line arguments override settings from the configuration file.
```

Reviewer 3: and here, I would add an admonition

    ```

    sends queries for `*.local` domains to `192.168.0.1:53`. Other queries are sent to `8.8.8.8:53`.

- A configuration like:

    ```none
    8.8.8.8:53
    [/host.com/]1.1.1.1:53
    [/maps.host.com/]#
    ```

    sends queries for `*.host.com` to `1.1.1.1:53` except for `*.maps.host.com` which are sent to `8.8.8.8:53` along with all other queries.

- A configuration like:

    ```none
    8.8.8.8:53
    [/host.com/]1.1.1.1:53
    [/*.host.com/]2.2.2.2:53
    ```

    sends queries for `*.host.com` to `2.2.2.2:53` except for `host.com` queries, those are sent to `1.1.1.1:53`, but all other queries are sent to `8.8.8.8:53`.

- A configuration like:

    ```none
    8.8.8.8:53
    [/host.com/]1.1.1.1:53 2.2.2.2:53
    ```

    sends queries for `*.host.com` to `1.1.1.1:53` and `2.2.2.2:53`, but all other queries are sent to `8.8.8.8:53`.

### Loading upstreams from file {#upstreams-from-file}

Using specific upstreams for some domains is a common way to accelerate internet in China. For an example, see https://github.com/felixonmars/dnsmasq-china-list or any other of the many `dnsmasq` lists. These lists can be easily converted to a list for AdGuard Home:

```none

> Before:  server=/0-100.com/114.114.114.114

Reviewer 3: ```suggestion
Before: server=/0-100.com/114.114.114.114
```

> After:   [/0-100.com/]114.114.114.114

Reviewer 3: ```suggestion
After: [/0-100.com/]114.114.114.114
```

```

The problem with these lists is that they may be too large. In this case you may want to load them from a separate file instead of setting all upstreams in AdGuard Home settings. To do that, simply specify the path to a file with your list in the `upstream_dns_file` field of `AdGuardHome.yaml`.

:::warning

The file, just like the input in the web interface, currently doesn't accept internationalized domain names (e.g. `пример.рф` or `例子.中国`). As a workaround, convert them to Punycode (e.g. `xn--e1afmkfd.xn--p1ai` or `xn--fsqu00a.xn--fiqs8s` respectively). See [issue 2915][2915].

:::

### Specifying upstreams for reverse DNS {#upstreams-for-rdns}

Using the domain-specific upstream notation, you can specify dedicated upstream DNS servers for reverse DNS (rDNS) requests. If you want **all** your `PTR` queries with ARPA domain to be redirected to `192.168.8.8`:

> 1. Enter the following into the “Upstream DNS servers” field on the “Settings

Reviewer 3: ```suggestion
1. Enter the following into the *Upstream DNS servers* field on the *Settings*
```

>     → DNS settings” page:

Reviewer 3: ```suggestion
    → *DNS settings* page:
```

    ```none
    [/in-addr.arpa/]192.168.8.8
    [/ip6.arpa/]192.168.8.8
    ```

2. Enter the following into the “Private reverse DNS servers” field on the
    same page below the previous field:

    ```none
    192.168.8.8
    ```

    There is no need to use the domain-specific notation here, unless you want to redirect requests for different private ranges to different upstream servers.

:::note

All upstreams for private ranges **must** go to the “Private reverse DNS servers” field **and not** the main “Upstream DNS servers” field. Entering something like `[/192.in-addr.arpa/]192.168.8.8` into the main field will have no effect.

:::

> Read below for more details.

Reviewer 3: this sentence can be omitted

Upon the first execution, a file named `AdGuardHome.yaml` will be created, with default values written in it. You can modify the file while your AdGuard Home service is not running. Otherwise, any changes to the file will be lost because the running program will overwrite them.

Settings are stored in [YAML][yaml] format, possible parameters that you can configure are listed below.

[yaml]: https://yaml.org/

### `http` {#http}

The `http` object configures Web interface.

- `address`: Web interface IP address with port to listen on.

- `session_ttl`: Web session TTL is a time duration in a human-readable format. The web user will stay signed in for this amount of time.

- `doh`: DNS-over-HTTPS configuration. It has the following properties:

    - `routes`: List of HTTP route patterns for DoH requests. Default routes
      are:

        - `GET /dns-query`

        - `POST /dns-query`

        - `GET /dns-query/{ClientID}`

        - `POST /dns-query/{ClientID}`

    - `insecure_enabled`: If `true`, allow DoH queries via unencrypted HTTP, for example to use with reverse proxies.

- `pprof`: Profiling HTTP handler configuration. See section [Profiling with pprof](#pprof).

>     - `enabled`:  Whether pprof is enabled or not.

Reviewer 3: ```suggestion
    - `enabled`: Whether pprof is enabled or not.
```

### `http_proxy` {#http_proxy}

Proxy URL for HTTP client. Supports `http`, `https` and `socks5` schemes.

**Example:** `http://user:password@server:port/`

### `language` {#language}

UI language code.

**Example:** `en`

### `theme` {#theme}

The UI theme. Possible values:

- `auto`

- `dark`

- `light`

### `dns` {#dns}

The `dns` object configures the DNS server. It has the following properties:

- `bind_hosts`: IP addresses on which to serve DNS queries. For each network interface there can only be one IP address of each IP version.

    **Example:**

    ```yaml
    # Different network interfaces.
    'dns':
        'bind_hosts':
        - '127.0.0.1'
        - '192.168.1.1'
    ```

    ```yaml
    # Same network interface, different IP versions.
    'dns':
        'bind_hosts':
        - '127.0.0.1'
        - '::1'
    ```

    If you want your server to accept requests on all interfaces and using both IP versions, for example if you run a public server, put **one** item with the unspecified IP of any version:

    ```yaml
    'dns':
        'bind_hosts':
        - '0.0.0.0'
    ```

- `port`: DNS server port to listen on.

> - `anonymize_client_ip`: If true, anonymize clients' IP addresses in logs and stats.

Reviewer 3: please use the unicode apostrophe (’) here and below

- `cache_time`: Safe Browsing, Safe Search, and Parental Control cache TTL, in seconds.

- `max_goroutines`: Maximum number of parallel goroutines for processing incoming requests.

- `handle_ddr`: Handle [Discovery of Designated Resolvers (DDR)][DDR] requests.

- `ipset`: List of domain-ipset_name associations for adding IP addresses of the specified domain names to an ipset list. Syntax:

    ```none
    DOMAIN[,DOMAIN,…]/IPSET_NAME[,IPSET_NAME,…]
    ```

    IPv4 addresses are added to an ipset list with `ipv4` family; IPv6 addresses, to an `ipv6` ipset list. ipset list must exist.

    This setting is supported on Linux OSs only. This feature is similar to `--ipset` in dnsmasq.

- `ipset_file`: Same as `ipset`, but the rules are read from a file. If this property is set, property `ipset` is ignored. Comments (lines starting with `#`) are supported.

- `upstream_timeout`: The timeout for querying upstream servers. Zero value will be rewritten with default one which is `10s`.

- `serve_http3`: Enables DNS-over-HTTP/3 serving for DNS-over-HTTPS clients as well as for the web UI.

- `serve_plain_dns`: Enables plain DNS serving.

> :::note

Reviewer: Indent the admonition.

    - `nxdomain`: Respond with NXDOMAIN code.

    - `refused`: Respond with REFUSED code.

- `blocking_ipv4`: IP address to be returned for a blocked A request if `blocking_mode` is set to `custom_ip`.

- `blocking_ipv6`: IP address to be returned for a blocked AAAA request if `blocking_mode` is set to `custom_ip`.

- `blocked_response_ttl`: For how many seconds the clients should cache a filtered response. Low values are useful on LAN if you change filters very often, high values are useful to increase performance and save traffic.

- `max_http_size`: The `max_http_size` property defines the maximum size of the HTTP request for rulelists. To disable the limitation, set a large size, such as `1 TB`.

- `protection_disabled_until`: Timestamp until when the protection is disabled.

- `parental_block_host`: IP (or domain name) which is used to respond to DNS requests blocked by parental control.

- `safebrowsing_block_host`: IP (or domain name) which is used to respond to DNS requests blocked by safe-browsing.

- `parental_enabled`: Parental control-based DNS requests filtering.

- `safe_search`: Safe search settings section. It has the following properties:

>     - `enabled`: Enforcing "Safe search" option for search engines, when possible.

Reviewer 3: here and below:
options/features/tabs should appear in italics instead of quotes
also, write either “Enforcing *Safe search*” or “Enforcing the *Safe search* option”

- `safebrowsing_block_host`: IP (or domain name) which is used to respond to DNS requests blocked by safe-browsing.

> - `parental_enabled`: *Parental control* based DNS requests filtering.

Reviewer 3: ```suggestion
- `parental_enabled`: Filtering of DNS requests based on *Parental control*.

- `safe_search`: *Safe search* settings section. It has the following properties:

    - `enabled`: Enforcing the *Safe search* option for search engines, when possible.

    - `bing`: Enforcing the *Safe search* option for `bing` domains.

    - `duckduckgo`: Enforcing the *Safe search* option for `duckduckgo` domains.

    - `ecosia`: Enforcing the *Safe search* option for `ecosia` domains.

    - `google`: Enforcing the *Safe search* option for `google` domains.

    - `pixabay`: Enforcing the *Safe search* option for `pixabay` domains.

    - `yandex`: Enforcing the *Safe search* option for `yandex` domains.

    - `youtube`: Enforcing the *Safe search* option for `youtube` domains.

> - `safebrowsing_enabled`: Filtering of DNS requests based on the *Safe browsing*.

Reviewer 3: ```suggestion
- `safebrowsing_enabled`: Filtering of DNS requests based on *Safe browsing*.
```

List of filters. Each filter has the following properties:

- `enabled`: Current filter’s status (enabled/disabled).

- `url`: URL pointing to the filter contents (filtering rules).

> - `name`: Name of the filter. If it’s an adguard syntax filter it will get updated automatically, otherwise it stays unchanged.

Reviewer 3: ```suggestion
- `name`: The name of the filter. If it’s an AdGuard syntax filter, it will be updated automatically; otherwise, it will stay unchanged.
```
